tensorflow_addons/image/autoaugment.py

```diff
@@ -0,0 +1,644 @@
+# limitations under the License.
+# ==============================================================================
+"""AutoAugment and RandAugment policies for enhanced image preprocessing.
+
+AutoAugment Reference: https://arxiv.org/abs/1805.09501
+RandAugment Reference: https://arxiv.org/abs/1909.13719
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import inspect
+import math
+import tensorflow as tf
+
+from tensorflow_addons.image import rotate as rot
+from tensorflow_addons.image import translate
+from tensorflow_addons.image import transform
+
+# This signifies the max integer that the controller RNN could predict for the
+# augmentation scheme.
+_MAX_LEVEL = 10.0
+
+
+def policy_v0():
+    """Autoaugment policy that was used in AutoAugment Paper."""
+    # Each tuple is an augmentation operation of the form
+    # (operation, probability, magnitude). Each element in policy is a
+    # sub-policy that will be applied sequentially on the image.
+    policy = [
+        [("Equalize", 0.8, 1), ("ShearY", 0.8, 4)],
+        [("Color", 0.4, 9), ("Equalize", 0.6, 3)],
+        [("Color", 0.4, 1), ("Rotate", 0.6, 8)],
+        [("Solarize", 0.8, 3), ("Equalize", 0.4, 7)],
+        [("Solarize", 0.4, 2), ("Solarize", 0.6, 2)],
+        [("Color", 0.2, 0), ("Equalize", 0.8, 8)],
+        [("Equalize", 0.4, 8), ("SolarizeAdd", 0.8, 3)],
+        [("ShearX", 0.2, 9), ("Rotate", 0.6, 8)],
+        [("Color", 0.6, 1), ("Equalize", 1.0, 2)],
+        [("Invert", 0.4, 9), ("Rotate", 0.6, 0)],
+        [("Equalize", 1.0, 9), ("ShearY", 0.6, 3)],
+        [("Color", 0.4, 7), ("Equalize", 0.6, 0)],
+        [("Posterize", 0.4, 6), ("AutoContrast", 0.4, 7)],
+        [("Solarize", 0.6, 8), ("Color", 0.6, 9)],
+        [("Solarize", 0.2, 4), ("Rotate", 0.8, 9)],
+        [("Rotate", 1.0, 7), ("TranslateY", 0.8, 9)],
+        [("ShearX", 0.0, 0), ("Solarize", 0.8, 4)],
+        [("ShearY", 0.8, 0), ("Color", 0.6, 4)],
+        [("Color", 1.0, 0), ("Rotate", 0.6, 2)],
+        [("Equalize", 0.8, 4), ("Equalize", 0.0, 8)],
+        [("Equalize", 1.0, 4), ("AutoContrast", 0.6, 2)],
+        [("ShearY", 0.4, 7), ("SolarizeAdd", 0.6, 7)],
+        [("Posterize", 0.8, 2), ("Solarize", 0.6, 10)],
+        [("Solarize", 0.6, 8), ("Equalize", 0.6, 1)],
+        [("Color", 0.8, 6), ("Rotate", 0.4, 5)],
+    ]
+    return policy
+
+
+def policy_vtest():
+    """Autoaugment test policy for debugging."""
+    # Each tuple is an augmentation operation of the form
+    # (operation, probability, magnitude). Each element in policy is a
+    # sub-policy that will be applied sequentially on the image.
+    policy = [
+        [("TranslateX", 1.0, 4), ("Equalize", 1.0, 10)],
+    ]
+    return policy
+
+
+def blend(image1, image2, factor):
+    """Blend image1 and image2 using 'factor'.
+
+  Factor can be above 0.0.  A value of 0.0 means only image1 is used.
```

Comment:
These ops could be separated into standalone image ops.

Reply:
I thought so too. But referring to [this](https://github.com/tensorflow/addons/issues/1226#issuecomment-595469029), would `AutoAugment` benefit from keeping it in the same module? Please expand on what that comment meant.

Reply:
Why single ops are not going in `tfa.image` namespace?

Reply:
My bad I'll put them in separate modules.

Reply:
@bhack can you please expand on how adding these ops to `tfa.image` would lead to better scheduling? I'm not aware how that works. How apparent would the overhead be if we were using keras preprocessing ops?

Reply:
I meant i.e. `solarize` is useful in `tensorflow_addons.image` so it will be not exclusive to `AutoAugment` and could be reused.
I think that `Keras preprocessing` it is another topic and it needs to find a way to coordinate with `tfa.image` and  `tf.image` over the implemented op. 
I think it is not ideal to spread image processing operations everywhere.

Reply:
I agree it is leading to a lot of boilerplate code. It would be great if Keras preprocessing could do that.

Reply:
It Is better that you open and try to review/merge another PR to add this image ops in and image ops in AutoAugment variant [remixmatch/ctaugment](https://github.com/google-research/remixmatch/blob/master/libml/ctaugment.py) in `tensorflow_addons.image`

Reply:
Sure, I'll just add one op in this PR, make new ones for other ops. Should make it easier to review and ultimately better code. Other discussions can be continued after ops are merged.